File: packages/protocol-types/src/sequence.ts
/**
 * 16-bit sequence validation and tracking with wraparound support.
 */

export const SEQUENCE_WINDOW_MAX = 32768;

/**
 * Checks if incoming sequence is a valid forward advance from latest sequence.
 * Enforces modular condition: `0 < (incoming - latest) mod 65536 < 32768`.
 */
export function isValidSequenceAdvance(latest: number, incoming: number): boolean {
  const diff = (incoming - latest) & 0xffff;
  return diff > 0 && diff < SEQUENCE_WINDOW_MAX;
}

/**
 * Tracks received packet sequences and drops out-of-order or duplicate packets.
 */
export class SequenceTracker {
  private _latest = 0;
  private _initialized = false;

  constructor(initialSequence?: number) {
    if (initialSequence !== undefined) {
      this._latest = initialSequence & 0xffff;
      this._initialized = true;
    }
  }

  public get latest(): number | null {
    return this._initialized ? this._latest : null;
  }

  public get isInitialized(): boolean {
    return this._initialized;
  }

  public reset(): void {
    this._latest = 0;
    this._initialized = false;
  }

  public check(incoming: number): boolean {
    if (!this._initialized) {
      return true;
    }
    return isValidSequenceAdvance(this._latest, incoming & 0xffff);
  }

  public checkAndUpdate(incoming: number): boolean {
    const seq = incoming & 0xffff;
    if (!this._initialized) {
      this._latest = seq;
      this._initialized = true;
      return true;
    }

    if (isValidSequenceAdvance(this._latest, seq)) {
      this._latest = seq;
      return true;
    }
    return false;
  }
}

/**
 * Monotonically increasing 16-bit sequence generator.
 */
export class SequenceGenerator {
  private _current: number;

  constructor(start = 1) {
    this._current = start & 0xffff;
  }

  public get current(): number {
    return this._current;
  }

  public next(): number {
    const seq = this._current;
    this._current = (this._current + 1) & 0xffff;
    return seq;
  }
}

File: packages/protocol/src/sequence.rs
//! 16-bit wrapping sequence filter and sequence generator.

/// Half the 16-bit sequence space (32768).
pub const SEQUENCE_WINDOW_MAX: u16 = 32768;

/// Evaluates whether `incoming` is a valid advance over `latest` according to modular arithmetic:
/// `0 < (incoming - latest) mod 65536 < 32768`.
#[inline(always)]
pub fn is_valid_sequence_advance(latest: u16, incoming: u16) -> bool {
    let diff = incoming.wrapping_sub(latest);
    diff > 0 && diff < SEQUENCE_WINDOW_MAX
}

/// Tracks the latest received sequence number and filters out duplicate or out-of-order frames.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SequenceTracker {
    latest: u16,
    initialized: bool,
}

/// Alias for `SequenceTracker`.
pub type SequenceFilter = SequenceTracker;

impl Default for SequenceTracker {
    #[inline(always)]
    fn default() -> Self {
        Self::new()
    }
}

impl SequenceTracker {
    /// Create a new uninitialized sequence tracker.
    /// The first packet received will always be accepted and initialize the tracker.
    #[inline(always)]
    pub const fn new() -> Self {
        Self {
            latest: 0,
            initialized: false,
        }
    }

    /// Create a sequence tracker initialized with a specific starting sequence.
    #[inline(always)]
    pub const fn with_initial(sequence: u16) -> Self {
        Self {
            latest: sequence,
            initialized: true,
        }
    }

    /// Returns the highest sequence number processed so far, if initialized.
    #[inline(always)]
    pub const fn latest(&self) -> Option<u16> {
        if self.initialized {
            Some(self.latest)
        } else {
            None
        }
    }

    /// Reset tracker state (e.g. on emergency reset or new session).
    #[inline(always)]
    pub fn reset(&mut self) {
        self.latest = 0;
        self.initialized = false;
    }

    /// Check if incoming sequence number is valid without updating state.
    #[inline(always)]
    pub fn check(&self, incoming: u16) -> bool {
        if !self.initialized {
            return true;
        }
        is_valid_sequence_advance(self.latest, incoming)
    }

    /// Process incoming sequence number.
    /// Returns `true` if accepted (and updates latest), or `false` if rejected (out of order or duplicate).
    #[inline(always)]
    pub fn check_and_update(&mut self, incoming: u16) -> bool {
        if !self.initialized {
            self.latest = incoming;
            self.initialized = true;
            return true;
        }

        if is_valid_sequence_advance(self.latest, incoming) {
            self.latest = incoming;
            true
        } else {
            false
        }
    }
}

/// Monotonic 16-bit sequence generator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SequenceGenerator {
    current: u16,
}

impl Default for SequenceGenerator {
    #[inline(always)]
    fn default() -> Self {
        Self::new()
    }
}

impl SequenceGenerator {
    /// Create a sequence generator starting at `1`.
    #[inline(always)]
    pub const fn new() -> Self {
        Self { current: 1 }
    }

    /// Create a sequence generator starting at a custom value.
    #[inline(always)]
    pub const fn with_start(start: u16) -> Self {
        Self { current: start }
    }

    /// Get current sequence without advancing.
    #[inline(always)]
    pub const fn current(&self) -> u16 {
        self.current
    }

    /// Yield next sequence number and advance monotonically with 16-bit wraparound.
    #[inline(always)]
    pub fn next_sequence(&mut self) -> u16 {
        let seq = self.current;
        self.current = self.current.wrapping_add(1);
        seq
    }
}
